source/simulator/core.cc

```diff
@@ -686,6 +686,13 @@ namespace aspect
     // wait if there is a thread that's still writing the statistics
     // object (set from the output_statistics() function)
     output_statistics_thread.join();
+
+    // Detect if we are being destructed because an uncaught exception has
+    // been triggered. If so, reset the TimerOutput class, because it tries to
+    // do MPI communication in its destructor otherwise, which can cause weird
+    // MPI errors or hangs.
+    if (std::uncaught_exception())
+      computing_timer.reset();
```

Comment:
`computing_timer` is a smart pointer, right? Doesn't resetting it cause the destructor of the object pointed to to run? Then what is the difference to before? The smart pointer would have gone out of scope at the end of this function anyway, so would have done the exact same operation.

What you would want to do (maybe?) is to actually leak the object, no?

Reply:
No, the type is ``TimerOutput`` and ``reset()`` does the right thing: https://github.com/dealii/dealii/blob/master/source/base/timer.cc#L681

It just empties the currently active sections, meaning the destructor will not do MPI communication. ;-)

Reply:
Ah, I see - not the kind of `reset()` I had had in mind.

Reply:
That now makes more sense to me. It will avoid *this processor* from initiating communication, though the others will then of course deadlock.

Would it not be nicer to make sure that exceptions can't propagate this far? We know where this particular exception comes from (and could just convert it to an `abort()` or `AssertNothrow`), though I agree that in the general case we have to deal with the situation where *any* kind of exception happens. Your solution looks reasonable in that case, though the deadlock is concerning.

Reply:
> Would it not be nicer to make sure that exceptions can't propagate this far? We know where this particular exception comes from (and could just convert it to an abort() or AssertNothrow),

Sure, we could change this one place, but any ``AssertThrow`` that triggers while we are inside a timing section will cause a similar issue (no reporting and a deadlock or reporting a confusing ``compress`` error that is unrelated). For that reason I prefer the current proposed solution.

At least on my machine (with this patch!) we will cleanly report the error and abort. Because the offending process is killed, MPI notices this and also kills the other processes (see below). This is not true without the patch.

I now get:
```
An error occurred in line <120> of file </ssd/aspect-git/source/particle/interpolator/cell_average.cc> in function (...)

The violated condition was: 
    non_empty_neighbors != 0
Additional information: 
    A cell and all of its neighbors do not contain any particles. The `cell average' interpolation scheme does not support this case.
--------------------------------------------------------

Aborting!
----------------------------------------------------
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code.. Per user-direction, the job has been aborted.
-------------------------------------------------------
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 6 in communicator MPI_COMM_WORLD 
with errorcode 59.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```

Reply:
OK, thanks for explaining.

Can you add a comment to this line that explains what you are doing and why, in particular why the `.reset()` call does and why that solves the problem? In essence a summary of the discussion above.

OK to merge if you add this!